### PR TITLE
fix(workers): prepare actor weights before rollout wakeup

### PR DESCRIPTION
**Pull request by gaohongkui**
## Summary

- Run the actor-side weight-preparation/offload phase, including the first non-merge LoRA base sync path, before resuming rollout weights in the colocated `naive` weight-sync path.
- Empty cached trainer-side allocations before waking rollout weights.
- Keep the async/disaggregated checkpoint-engine path unchanged.

## Motivation

For colocated trainer + rollout jobs with `free_cache_engine=True`, the old order resumed rollout weights before the actor engine had a chance to run its weight-sync preparation/offload phase:

```python
await self.rollout.resume(tags=["weights"])
per_tensor_param, peft_config = self.actor.engine.get_per_tensor_param(...)
```

In the FSDP/DTensor path, `get_per_tensor_param()` returns a lazy iterator; full tensors are materialized later when `BucketedWeightSender` consumes that iterator. The benefit of this PR is therefore not that all sync tensors are eagerly materialized before `resume()`. The important part is that `get_per_tensor_param()` performs actor-side setup such as `load_fsdp_model_to_gpu(...)`, state-dict / parameter iterator setup, and `offload_fsdp_model_to_cpu(...)`; this PR then clears cache before vLLM-Ascend wakes rollout weights.

That ordering avoids calling vLLM-Ascend `wake_up` while actor/FSDP memory and cache from the training side are still resident. In an internal Ascend/vLLM-Ascend 32-NPU Qwen3.5-35B-A3B GRPO smoke test, the old order failed during vLLM-Ascend `wake_up` with:

```text
aclrtMallocPhysical failed with acl error code: 207001(OOM: Out of Memory, allocation failed)
```

The reordered version completed the same 2-step smoke test, including validation and checkpoint save. The PR also keeps the first non-merge LoRA base sync on the same pre-resume ordering to avoid reintroducing the same memory overlap on that path.

## Duplicate-work check

I checked for overlapping work before opening this PR:

```bash
gh issue view 6225 --repo verl-project/verl --comments
gh pr list --repo verl-project/verl --state open --search "6225 in:body"
gh pr list --repo verl-project/verl --state open --search "update_weights wake_up OOM"
gh pr list --repo verl-project/verl --state open --search "rollout resume weights get_per_tensor_param"
gh pr list --repo verl-project/verl --state open --search "ActorRolloutRefWorker update_weights resume weights"
```

No open PR appears to address this exact ordering fix. #6225 discusses a broader colocated checkpoint-engine / batched IPC design; this PR is a smaller stability fix for the existing `naive` path.

## Tests

```bash
git diff --check
python3 -m py_compile verl/workers/engine_workers.py
uv venv --python 3.12
source .venv/bin/activate
uv pip install pre-commit hydra-core
pre-commit run --files verl/workers/engine_workers.py
```

`pre-commit run --files verl/workers/engine_workers.py` passed all configured hooks.

Runtime A/B smoke test on internal Ascend/vLLM-Ascend 32-NPU Qwen3.5-35B-A3B GRPO setup:

- patched order: completed 2/2 training steps, validation, and checkpoint save;
- old order: failed in `self.rollout.resume(tags=["weights"])` / vLLM-Ascend `wake_up` with CaMem OOM.

## AI-assisted disclosure

AI assistance (Codex) was used to prepare this draft PR. The submitting human should review every changed line and the test evidence before marking the PR ready for review.

### CODE REVIEW

**Comment by gaohongkui** [comment]
Thanks for the review. I agree this edge case should follow the same ordering.

Updated the PR so the non-merge LoRA first base-sync path also prepares `per_tensor_param_base` and clears cache before `rollout.resume(tags=["weights"])`.

Re-ran `pre-commit run --files verl/workers/engine_workers.py`; all hooks passed.

**Comment by wuxibin89** [comment]
I don't get why this reordering reduce peak memory?

**Comment by gaohongkui** [comment]
Thanks for asking. The wording in the PR description was a bit too terse: this does not reduce the steady-state memory footprint of the model. It reduces the transient peak in the colocated `naive` weight-sync path by changing which allocations overlap in time.

With the old order, the sequence is roughly:

```python
# rollout weights were previously put to sleep/released
await self.rollout.resume(tags=["weights"])   # vLLM-Ascend wake_up allocates CaMem for rollout weights
per_tensor_param, peft_config = self.actor.engine.get_per_tensor_param(...)
                                                   # actor/FSDP side then materializes tensors for weight sync
```

So during `update_weights`, rollout weight memory is already restored while the trainer side still needs temporary/export tensors for `get_per_tensor_param(...)` and the following transfer. In the reordered path, we first prepare the actor-side tensors, clear the trainer-side cache, and only then wake the rollout weights:

```python
per_tensor_param, peft_config = self.actor.engine.get_per_tensor_param(...)
aggressive_empty_cache(force_sync=True)
await self.rollout.resume(tags=["weights"])
```

I re-ran a controlled A/B smoke test on the same internal Ascend/vLLM-Ascend 32-NPU Qwen3.5-35B-A3B GRPO setup with the same image/config, changing only this ordering:

- patched order: completed 2/2 training steps, including validation and checkpoint save. The logs show `training/global_step:1`, `timing_s/update_weights:60.72`, then `training/global_step:2`, `timing_s/update_weights:60.01`, and final `Training Progress: 100%|...| 2/2`.
- old order: failed before completing step 1 during vLLM-Ascend `wake_up` from `resume(tags=["weights"])` with:

```text
aclrtMallocPhysical failed with acl error code: 207001(OOM: Out of Memory, allocation failed)
/opt/src/vllm-ascend/csrc/camem_allocator.cpp:67

vllm_ascend/device_allocator/camem.py", line 214, in wake_up
vllm_ascend/worker/worker.py", line 215, in wake_up
vllm/v1/engine/core.py", line 692, in wake_up
```

That is the peak this PR is trying to avoid: not by changing the final set of model/rollout tensors, but by avoiding the overlap between vLLM rollout weight wake-up allocations and actor-side temporary weight export/sync allocations. The non-merge LoRA first base-sync path is handled the same way for the same reason.

**Comment by wuxibin89** [comment]
> In the reordered path, we first prepare the actor-side tensors, clear the trainer-side cache, and only then wake the rollout weights:

But the `get_per_tensor_param` just return a python generator, it actually not materialize full tensor when calling.
https://github.com/verl-project/verl/blob/release/v0.8.0/verl/workers/engine/fsdp/transformer_impl.py#L838-L846 

The generator materialize full tensor when consumed by BucketedWeightSender:
https://github.com/verl-project/verl/blob/release/v0.8.0/verl/workers/rollout/vllm_rollout/bucketed_weight_transfer.py#L120

**Comment by gaohongkui** [comment]
You're right, thanks for pointing this out. My previous explanation was imprecise.

For the FSDP/DTensor path, `get_per_tensor_param()` returns a lazy iterator, and the `full_tensor()` materialization happens later when `BucketedWeightSender` consumes the iterator. So this patch is not reducing the peak by pre-materializing all sync tensors before `rollout.resume(tags=["weights"])`.

The part that still matters before `resume()` is that `get_per_tensor_param()` runs the actor-side preparation phase: `load_fsdp_model_to_gpu(...)`, state-dict / parameter iterator setup, then `offload_fsdp_model_to_cpu(...)`, followed by `aggressive_empty_cache(force_sync=True)` in this PR. In the old order, vLLM-Ascend `wake_up` was called before this actor-side offload/cache cleanup, so rollout weight allocation could overlap with actor/FSDP memory still resident after training. In the patched order, the actor-side setup and offload/cache cleanup happen first; then rollout weights are woken up. The later tensor materialization during `update_weights()` is still bucketed by `BucketedWeightSender` / `update_weights_bucket_megabytes`.

This also matches the A/B failure point we observed: the old order failed inside vLLM-Ascend `wake_up` / CaMem allocation before step 1 completed, while the patched order completed the 2-step smoke. I will update the PR description/comment wording to avoid implying that `get_per_tensor_param()` eagerly materializes all tensors.

File: verl/workers/engine_workers.py [hunk 1]
```diff
@@ -700,28 +700,38 @@ async def update_weights(self, global_steps: int = None, mode: str = "auto"):
...
-        log_gpu_memory_usage("After resume weights", logger=logger)
-
-        # 2. determine if we need a base weight sync (adapter path only)
+        # 1. Prepare actor weights before waking rollout weights. This keeps
+        # rollout memory from overlapping with temporary trainer-side tensors
+        # created while exporting parameters for weight sync.
         per_tensor_param, peft_config = self.actor.engine.get_per_tensor_param(
             layered_summon=self.layered_summon, base_sync_done=True
         )
+        aggressive_empty_cache(force_sync=True)
```
Comment (wuxibin89) [inline verl/workers/engine_workers.py:712]: I think which actually take effect is aggressive_empty_cache, could you try just adding this line before #L702?
```python
aggressive_empty_cache(force_sync=True)
set_expandable_segments(False)
```
